yt_idefix/data_structures.py

```diff
@@ -427,15 +427,175 @@ def _get_time(self):
+                    uo["density_unit"] = (
+                        (uo["magnetic_unit"] / uo["velocity_unit"]) ** 2 / 4.0 / np.pi
+                    )
+                else:
+                    uo["density_unit"] = (
+                        (uo["magnetic_unit"] * uo["time_unit"]) ** 3 / uo["mass_unit"]
+                    ) ** 2 / (4.0 * np.pi) ** 3
+
+        unit_deps = {
+            "length_unit": _length,
+            "mass_unit": _mass,
+            "time_unit": _time,
+            "velocity_unit": _velocity,
+            "density_unit": _density,
+        }
+
+        def cal_unit(unit):
+            """Derived units by recursion"""
+            try:
+                unit_deps[unit]()
+            except KeyError as err:
+                missed = err.args[0]
+                cal_unit(missed)
+                cal_unit(unit)
+
+        # Derive base units, the order doesn't matter
+        cal_unit("mass_unit")
+        cal_unit("length_unit")
+        cal_unit("time_unit")
+
+        self.mass_unit = uo["mass_unit"]
+        self.length_unit = uo["length_unit"]
+        self.time_unit = uo["time_unit"]
+
+        self.velocity_unit = self.length_unit / self.time_unit
+        self.density_unit = self.mass_unit / self.length_unit ** 3
+        self.magnetic_unit = (
+            np.sqrt(4.0 * np.pi * self.density_unit) * self.velocity_unit
+        )
         self.magnetic_unit.convert_to_units("gauss")
+        self.temperature_unit = self.quan(1.0, "K")
+
+    invalid_unit_combinations = [
+        {"magnetic_unit", "velocity_unit", "density_unit"},
+        {"velocity_unit", "time_unit", "length_unit"},
+        {"density_unit", "length_unit", "mass_unit"},
+    ]
+
+    default_units = {
+        "length_unit": "cm",
+        "time_unit": "s",
+        "mass_unit": "g",
+        "velocity_unit": "cm/s",
+        "magnetic_unit": "gauss",
+        "temperature_unit": "K",
+        # this is the one difference with Dataset.default_units:
+        # we accept density_unit as a valid override
+        "density_unit": "g/cm**3",
+    }
+
+    @classmethod
+    def _validate_units_override_keys(cls, units_override):
+        """Check that units in units_override are able to derive three base units:
+        mass, length and time
+        """
+
+        # YT supports overriding other normalisations, this method ensures consistency
+        # between supplied 'units_override' items and principles in PLUTO.
+
+        # PLUTO's normalisations/units have 3 degrees of freedom. Therefore, any combinations
+        # are valid other than the three cases listed explicitly.
+
+        if "temperature_unit" in units_override:
+            raise ValueError(
+                "Temperature is not allowed in units_override,"
+                "since it's always in Kelvin in PLUTO"
+            )
+
+        # Three units are enough for deriving others, more will cause conflict
+        if len(units_override) > 3:
```

Comment:
Are you sure that's correct ? I would assume that overriding magnetic as well as length, mass and time is fine ?

Reply:
Since we have only three base units here: length, mass and time, just like solving a system of equations containing three unknowns, three equations are enough, or three base/compound units here.

So an additional magnetic might conflict with that derived with length, mass and time. "more will _likely_ cause conflict" is more precise I think.